Fixes DISCO-2605: Replace template parameters in AMP suggestion URLs.

linabutler opened the pull request:
Some AMP suggestion URLs contain template parameters, like `%YYYYMMDDHH%`, which we must "cook" with real values before returning the suggestion. This means that an AMP suggestion's URL can vary depending on when it's shown.

When deduping suggestions, though, we want to treat all variations of these URLs as equivalent, and ignore the replaced parameters. Otherwise, the address bar might show an AMP suggestion twice, as on Desktop in bug 1735894: once from history, with one set of parameters; and once from Suggest, with a different set of parameters.

To address both of these needs, this commit adds:

* A `SuggestionUrl` type that holds both the "cooked" URL string, and the locations of replaced template parameters within that string.
* A `suggestion_url_matches` function that compares a `SuggestionUrl` with a URL string, to determine if they're identical except for their replaced parameters.

/cc @0c0w3

### Pull Request checklist ###
<!-- Before submitting the PR, please address each item -->
- **Breaking changes**:  This PR follows our [breaking change policy](https://github.com/mozilla/application-services/blob/main/docs/howtos/breaking-changes.md)
  - [x] This PR follows the breaking change policy:
     - This PR has no breaking API changes, or
     - There are corresponding PRs for our consumer applications that resolve the breaking changes and have been approved
- [x] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/cut-a-new-release.md) after merging.
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes a changelog entry in [CHANGELOG.md](../CHANGELOG.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [x] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

[Branch builds](https://github.com/mozilla/application-services/blob/main/docs/howtos/branch-builds.md): add `[firefox-android: branch-name]` to the PR title.

0c0w3 commented:
Thanks @linabutler. Desktop stores the original URL alongside the cooked URL (thanks for that term) in the suggestion object so that we can quickly and easily store and look up dismissed suggestions. It looks like clients can use the replacements concept in this PR to produce uncooked URLs to accomplish the same thing, but that's not quite as convenient. wdyt?

linabutler commented on components/suggest/src/suggest.udl:
```diff
@@ -8,35 +8,46 @@ typedef extern RemoteSettingsConfig;
     "Wikipedia",
 };
 
+dictionary SuggestionUrl {
+    string string;
+    sequence<SuggestionUrlReplacement> replacements;
+};
```
This interface is a little awkward because of mozilla/uniffi-rs#1372. At first, I modeled `SuggestionUrl` as an object, like:

```webidl
interface SuggestionUrl {
    string string();
    boolean matches([ByRef] string other_url_string);
};
```

But `Suggestion` is an enum with associated data, and those can't have object fields yet. Even if they could, though, `SuggestionUrl` would need an explicit destructor in Kotlin (mozilla/uniffi-rs#1394).

Another idea I considered was turning `Suggestion` into an "enum-like" object with optional-returning getters for each variant, like this:

```webidl
interface Suggestion {
    AmpSuggestion? as_amp_suggestion();
    WikipediaSuggestion? as_wikipedia_suggestion();
};

dictionary AmpSuggestion {
    string title;
    SuggestionUrl url;
    sequence<u8>? icon;
    // ...
};

dictionary WikipediaSuggestion {
    string title;
    SuggestionUrl url;
    // ...
};
```

But that makes `Suggestion` feel clunkier, and forces explicit destructors for `Suggestion` _and_ `SuggestionUrl` in Kotlin.

Also, UniFFI dictionaries can't have methods, so that's why `suggestion_url_matches` is a free function instead of a method on `SuggestionUrl`.

/cc @bendk @mhammond Do you folks have any suggestions (😅) for other things to try, or does this interface look okay to you, given UniFFI's restrictions on it?

bendk replied:
I can't really think of anything better.  This makes me want to address 1372 and let users configure if they want to treat their interface as a resource or not.

linabutler commented:
Thanks @0c0w3!

> Desktop stores the original URL alongside the cooked URL (thanks for that term) in the suggestion object so that we can quickly and easily store and look up dismissed suggestions. It looks like clients can use the replacements concept in this PR to produce uncooked URLs to accomplish the same thing, but that's not quite as convenient. wdyt?

Ooh, thanks for mentioning dismissed suggestions—I was so focused on equivalence that I didn't think about that being another case where we'd need the original "raw" URL, oops! 😬

I wonder, do you think we could get away with passing just the raw and cooked URLs back, and drop the "replacements" concept (and not worry about implementing equivalence in the Rust component?) altogether?

It would definitely simplify `SuggestionUrl`:

```webidl
interface SuggestionUrl {
    string cooked;
    string raw;
};
```

0c0w3 commented:
Yes, that's how I would do it. Equivalence testing is still useful though, especially for deduping as you mentioned. Since it's the Rust component that will be cooking these URLs -- it's the thing that knows which substrings need replacing and how to replace them -- IMO it should provide the equivalence test too. The `suggestion_url_matches` free function looks good to me, and desktop implements a similar function by storing another property on each AMP suggestion object, `urlTimestampIndex` (the timestamp length is assumed to be constant).

That approach may or may not be scalable to more or different types of templates, I'm not sure. Actually, a related thing that came up recently is UTM params. Only a couple suggestion types use them and they never change, so we've lucked out so far and we don't have any special handling for these at all, but if they ever do change, like timestamp templates do, we'd need to handle them too.

linabutler commented:
Great, thanks so much, @0c0w3! Here's what I changed in the latest push:

* Dropped `SuggestionUrl` and `SuggestionUrlReplacement` 🗑️
* Added `raw_url` and `raw_click_url` fields for `Amp` suggestions. I made these optional at first, and only set when the raw URL != the cooked URL, but then changed them to be set always. Please let me know if you'd prefer it the first way, and I can drop the second commit 😊
* Renamed `suggestion_url_matches` to `raw_suggestion_url_matches`. Functionally, it works the same way, and compares the non-template segments of the raw and cooked URLs.

`raw_suggestion_url_matches` makes the same assumption as Desktop that the timestamp length is constant, and that it's the only param. We've been trying to avoid taking a dependency on the `regex` crate, since it impacts our binary size by quite a bit—but I think we can still extend `raw_suggestion_url_matches` to do variable-length matching if and when we need it for other params.

linabutler commented on components/suggest/src/suggest.udl:
```diff
@@ -26,13 +27,15 @@ interface Suggestion {
     Amp(
         string title,
         string url,
+        string raw_url,
```
note: This is a source breaking change for Swift consumers (which we don't have, yet! 🤫), but not Kotlin or JS.

0c0w3 reviewed (approved):
Thanks, lgtm! The desktop impl has a couple optimizations over this I thought I'd mention in case you're interested.

First, it compares cooked URLs to other cooked URLs, so right off the bat it can return false if the two URLs aren't the same length, short circuiting the later substring operations. (A side benefit of comparing cooked URLs is it makes the raw URL less a part of the interface and more of an internal detail, although consumers end up needing the raw URL anyway to handle dismissals.)

Second, when it initially replaces a suggestion's templates, it also stores their indexes in the suggestion object. That lets the matching function avoid another full-string search for templates, keeping the function on the order of the number of segments in the URLs instead of their length.

Are these optimizations important? Maybe not. At most N calls to the matching function happen per keystroke, N = the number of history results alongside the Suggest result, and for desktop N is always a very small number, less than 10 by default.

linabutler commented:
That makes perfect sense! I think my original implementation was closer to Desktop's—`SuggestionUrl` held the cooked URL, and the positions of any replacements in it.

After thinking about it, though, I wasn't sure if the benefit of avoiding the linear scan was enough to overcome the other costs. The biggest one is that exposing `SuggestionUrl` as a dictionary would add a UniFFI serialization toll, which we'd pay twice: once when returning the `Suggestion` from `query`, and a second time when calling the (since-renamed) `suggestion_url_matches` function.

In addition to serialization, there's a bit of additional complexity to validate that the replacement positions are valid. Because `SuggestionUrl` is a dictionary, all its fields become part of our component's public interface—so, even though we don't expect our consumers to ever create their own `SuggestionUrl`s, we'd still need to handle potentially invalid positions.

I think the API that I'd really love to have—and that would also solve the issue of exposing replacement positions over the FFI—is a `SuggestionUrl` object, rather than a dictionary. The object could encapsulate the "raw" URL and template replacement positions, so that consumers wouldn't need to know about them at all. IOW, something like:

```webidl
interface SuggestionUrl {
    // Returns the URL as a "cooked" string, with all template strings replaced
    // with real values.
    string string();

    // Returns `true` if this URL is equivalent to a "cooked" `other_url`.
    // Internally, this would compare this URL's cooked string to `other_url`,
    // then compare this URL's "raw" string to `other_url`, ignoring any
    // replaced templates.
    boolean matches(string other_url_string);
};

[Enum]
interface Suggestion {
    Amp(
        string title,
        SuggestionUrl url,
        SuggestionUrl click_url
        /* ...Other fields elided... */
    );
    Wikipedia(
        string title,
        SuggestionUrl url
        /* ...Other fields elided... */
    );
};

interface SuggestStore {
    /* ...Other store methods elided... */

    // Marks a suggestion URL as dismissed. Internally, this would get the
    // "raw" URL from the `url`, and mark that as dismissed.
    void dismiss(SuggestionUrl url);
};
```

...But I think we'd need to wait for mozilla/uniffi-rs#1372 (and mozilla/uniffi-rs#1394) to have that.

0c0w3 commented:
Thanks for your reply! That interface does look nice. I hadn't considered that the suggest component would keep track of dismissed suggestions too -- I was thinking each platform would need to do it -- but that makes perfect sense and it would certainly make my job easier. :-)

ncloudioj commented on components/suggest/src/suggestion.rs:
```diff
@@ -31,3 +46,184 @@ impl Suggestion {
+/// Determines whether a "raw" sponsored suggestion URL is equivalent to a
+/// "cooked" URL. The two URLs are equivalent if they are identical except for
+/// their replaced template parameters, which can be different.
+pub fn raw_suggestion_url_matches(raw_url: &str, cooked_url: &str) -> bool {
```
Likewise, if we leverage the invariant of at-most-1-template-for-each-url, this can be simplified as well.

ncloudioj commented on components/suggest/src/suggestion.rs:
```diff
@@ -3,19 +3,34 @@
     Amp {
         title: String,
         url: String,
+        raw_url: String,
```
Worth noting that there can be no template in the raw urls for certain advertisers, meaning that `url` and `raw_url` will be the same for those. Though the vast majority of suggestions do have that, so I think we can live with that tiny overhead for now.

linabutler replied:
Definitely! I made these `Optional<String>` at first, and only passed them if `url` and `raw_url` are different—but I think always passing them makes things a bit easier for consumers, at the cost of some overhead for us. Otherwise, our consumers would need to do something like `suggestion.rawUrl ?? suggestion.url` to get at the raw URL.

ncloudioj reviewed (approved):
Looks good, r+wc.

linabutler commented on components/suggest/src/suggestion.rs:
```diff
@@ -31,3 +46,184 @@ impl Suggestion {
+    let now = Local::now();
+    let mut cooked_url = String::new();
+    let mut last_index = 0;
+    for (index, _) in raw_url.match_indices(TIMESTAMP_TEMPLATE) {
```
Replacing just the first occurrence sounds good to me!

> If the partner breaks that contract by including more than one template, we should reject it at the ingestion phase and let them fix that. WDYT?

Agreed—just to make sure we're on the same page, though, I think it would be valuable to do this when we ingest into Remote Settings, not when we ingest from Remote Settings into our SQLite database.

ncloudioj replied:
Yes, I meant enforcing that at the ingestion phase before loading data into Remote Settings. Oh, right. We also have another ingestion here from RS to SQLite, but that's not what was referring to, sorry about that. 😅

linabutler replied:
No worries, I figured that's what you meant, and wanted to double-check!
